File: lib/models/checklist_group.dart
import 'checklist_item.dart';

class ChecklistGroup {
  final String id;
  final String title;
  final List<ChecklistItem> items;
  final String? inviteCode;
  final String? ownerToken;
  final Map<String, String> memberPermissions; // memberId -> "READ_WRITE" or "READ_ONLY"
  final Map<String, String> memberNicknames; // memberId -> nickname
  final List<String> kickedMembers;

  ChecklistGroup({
    required this.id,
    required this.title,
    this.items = const [],
    this.inviteCode,
    this.ownerToken,
    this.memberPermissions = const {},
    this.memberNicknames = const {},
    this.kickedMembers = const [],
  });

  factory ChecklistGroup.fromJson(Map<String, dynamic> json) {
    return ChecklistGroup(
      id: json['id'] as String,
      title: json['title'] as String,
      items: (json['items'] as List<dynamic>?)
              ?.map((e) => ChecklistItem.fromJson(e as Map<String, dynamic>))
              .toList() ??
          [],
      inviteCode: json['inviteCode'] as String?,
      ownerToken: json['ownerToken'] as String?,
      memberPermissions: Map<String, String>.from(json['memberPermissions'] ?? {}),
      memberNicknames: Map<String, String>.from(json['memberNicknames'] ?? {}),
      kickedMembers: List<String>.from(json['kickedMembers'] ?? []),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'items': items.map((e) => e.toJson()).toList(),
      'inviteCode': inviteCode,
      'ownerToken': ownerToken,
      'memberPermissions': memberPermissions,
      'memberNicknames': memberNicknames,
      'kickedMembers': kickedMembers,
    };
  }

  ChecklistGroup copyWith({
    String? id,
    String? title,
    List<ChecklistItem>? items,
    String? inviteCode,
    String? ownerToken,
    Map<String, String>? memberPermissions,
    Map<String, String>? memberNicknames,
    List<String>? kickedMembers,
  }) {
    return ChecklistGroup(
      id: id ?? this.id,
      title: title ?? this.title,
      items: items ?? this.items,
      inviteCode: inviteCode ?? this.inviteCode,
      ownerToken: ownerToken ?? this.ownerToken,
      memberPermissions: memberPermissions ?? this.memberPermissions,
      memberNicknames: memberNicknames ?? this.memberNicknames,
      kickedMembers: kickedMembers ?? this.kickedMembers,
    );
  }
}

File: server-spring/src/main/java/com/hatggo/server/ChecklistGroup.java
package com.hatggo.server;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class ChecklistGroup {
    private Long id;
    private String title;
    
    @JsonProperty("invite_code")
    private String inviteCode;
    
    private String type; // "DAILY_ROUTINE" or "DEADLINE_BOUND"
    
    @JsonProperty("start_date")
    private String startDate; // ISO 8601 string
    
    @JsonProperty("end_date")
    private String endDate; // ISO 8601 string
    
    @JsonProperty("owner_token")
    private String ownerToken;
    
    private Map<String, String> memberPermissions = new java.util.concurrent.ConcurrentHashMap<>(); // memberId -> "READ_WRITE" or "READ_ONLY"
    private Set<String> kickedMembers = java.util.Collections.newSetFromMap(new java.util.concurrent.ConcurrentHashMap<String, Boolean>()); // memberId set
    private Map<String, String> memberNicknames = new java.util.concurrent.ConcurrentHashMap<>(); // memberId -> nickname
    private List<ChecklistItem> items = new ArrayList<>();

    public ChecklistGroup() {}

    public ChecklistGroup(Long id, String title, String inviteCode, String type, String startDate, String endDate) {
        this.id = id;
        this.title = title;
        this.inviteCode = inviteCode;
        this.type = type != null ? type : "DAILY_ROUTINE";
        this.startDate = startDate;
        this.endDate = endDate;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getInviteCode() {
        return inviteCode;
    }

    public void setInviteCode(String inviteCode) {
        this.inviteCode = inviteCode;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getStartDate() {
        return startDate;
    }

    public void setStartDate(String startDate) {
        this.startDate = startDate;
    }

    public String getEndDate() {
        return endDate;
    }

    public void setEndDate(String endDate) {
        this.endDate = endDate;
    }

    public List<ChecklistItem> getItems() {
        return items;
    }

    public void setItems(List<ChecklistItem> items) {
        this.items = items;
    }

    public String getOwnerToken() {
        return ownerToken;
    }

    public void setOwnerToken(String ownerToken) {
        this.ownerToken = ownerToken;
    }

    public Map<String, String> getMemberPermissions() {
        return memberPermissions;
    }

    public void setMemberPermissions(Map<String, String> memberPermissions) {
        this.memberPermissions = memberPermissions;
    }

    public Set<String> getKickedMembers() {
        return kickedMembers;
    }

    public void setKickedMembers(Set<String> kickedMembers) {
        this.kickedMembers = kickedMembers;
    }

    public Map<String, String> getMemberNicknames() {
        return memberNicknames;
    }

    public void setMemberNicknames(Map<String, String> memberNicknames) {
        this.memberNicknames = memberNicknames;
    }
}
